### PR TITLE
feat: introduce per-sync configurations

### PR DESCRIPTION
**Pull request by james-milligan**
## This PR

- introduces the `SyncProviderConfig` object and associated start flags to pass configuration to specific sync providers

### Related Issues

https://github.com/open-feature/flagd/issues/405

### CODE REVIEW

File: pkg/runtime/from_config.go
Comment (james-milligan) [inline pkg/runtime/from_config.go:121]: Ive opted to not introduce unit tests for these contractors, although happy to if code coverage is required

File: cmd/start.go
Comment (skyerus) [inline cmd/start.go:115]: Is there a more appropriate name for this var?

File: pkg/runtime/from_config.go
Comment (skyerus) [inline pkg/runtime/from_config.go:84]:
```suggestion
			rtLogger.Debug(fmt.Sprintf("using filepath sync-provider for: %s", syncProvider.URI))
```

File: pkg/runtime/from_config.go
Comment (skyerus) [inline pkg/runtime/from_config.go:96]:
```suggestion
			rtLogger.Debug(fmt.Sprintf("using remote sync-provider for: %s", syncProvider.URI))
```

**Review by beeme1mr — COMMENTED** [review]
I wasn't able to get the examples working. I built flagd using your branch and confirmed the config example flags were in the expected location.

```sh
./flagd start --sync-providers=\[{\"uri\":\"config/samples/example_flags.json\"\,\"provider\":\"file\"}\]

                 ______   __       ________   _______    ______      
                /_____/\ /_/\     /_______/\ /______/\  /_____/\     
                \::::_\/_\:\ \    \::: _  \ \\::::__\/__\:::_ \ \    
                 \:\/___/\\:\ \    \::(_)  \ \\:\ /____/\\:\ \ \ \   
                  \:::._\/ \:\ \____\:: __  \ \\:\\_  _\/ \:\ \ \ \  
                   \:\ \    \:\/___/\\:.\ \  \ \\:\_\ \ \  \:\/.:| | 
                    \_\/     \_____\/ \__\/\__\/ \_____\/   \____/_/ 

2023-03-01T14:01:42.843-0500    info    cmd/start.go:104        flagd version: dev (5ecdb19), built at: 2023-03-01T14:00:02Z    {"component": "start"}
2023-03-01T14:01:42.843-0500    fatal   cmd/start.go:144        no sync implementation set      {"component": "start"}
github.com/open-feature/flagd/cmd.glob..func1
        /home/beemer/code/openfeature/flagd/cmd/start.go:144
github.com/spf13/cobra.(*Command).execute
        /home/beemer/go/pkg/mod/github.com/spf13/cobra@v1.6.1/command.go:920
github.com/spf13/cobra.(*Command).ExecuteC
        /home/beemer/go/pkg/mod/github.com/spf13/cobra@v1.6.1/command.go:1044
github.com/spf13/cobra.(*Command).Execute
        /home/beemer/go/pkg/mod/github.com/spf13/cobra@v1.6.1/command.go:968
github.com/open-feature/flagd/cmd.Execute
        /home/beemer/code/openfeature/flagd/cmd/root.go:37
main.main
        /home/beemer/code/openfeature/flagd/main.go:30
runtime.main
        /home/beemer/sdk/go1.19.6/src/runtime/proc.go:250
```

File: docs/configuration/flagd_start.md
Comment (james-milligan) [reply docs/configuration/flagd_start.md]: done

File: docs/configuration/flagd_start.md
Comment (james-milligan) [reply docs/configuration/flagd_start.md]: resolved

File: pkg/sync/file/filepath_sync.go
Comment (Kavindu-Dodan) [inline pkg/sync/file/filepath_sync.go:22]: Do we use this Configuration inside ISync implementation?  🤔

File: pkg/sync/isync.go
Comment (Kavindu-Dodan) [inline pkg/sync/isync.go:61]: should we focus on a per-provider configuration rather than a single one ? 🤔  For example, if we now introduce grpc specific configuration, then this get passed to all sync provider types

**Comment by Kavindu-Dodan** [comment]
General observation while attempting to introduce grpc tls connection parameters. My view is that,

- we need provider-specific configuration so that it is easy to decode related configurations and build providers based on the specific configuration
- avoid parsing the configuration down to the implementation level. we should either simply use struct initialization or use a NewSyncX(configuration) mechanism.

**Comment by beeme1mr** [comment]
> General observation while attempting to introduce grpc tls connection parameters. My view is that,
> 
> * we need provider-specific configuration so that it is easy to decode related configurations and build providers based on the specific configuration
> * avoid parsing the configuration down to the implementation level. we should either simply use struct initialization or use a NewSyncX(configuration) mechanism.

The seems reasonable to me.

**Comment by toddbaert** [comment]
> General observation while attempting to introduce grpc tls connection parameters. My view is that,
> 
> * we need provider-specific configuration so that it is easy to decode related configurations and build providers based on the specific configuration
> * avoid parsing the configuration down to the implementation level. we should either simply use struct initialization or use a NewSyncX(configuration) mechanism.

Are you saying we should probably merge this (or something like it) before adding back TLS support to syncs?

File: pkg/sync/isync.go
Comment (james-milligan) [reply pkg/sync/isync.go:61]: You're right that the field will be passed to all provider types, but it would be a nil reference of the type. IMO I don't see any issue with using a shared data structure - however, if we feel there is a need to introduce one im happy to make these changes. 

Introducing per-sync configuration objects will also add complexity to the passing / parsing of the configurations as startup commands, which is the main argument for using a single shared type (unmarshal directly into a single shared slice, rather than having a []typeXYZ{} slice to accommodate every new data type, or storing all configurations as interfaces and inferring their types on startup)

**Comment by james-milligan** [comment]
> avoid parsing the configuration down to the implementation level. we should either simply use struct initialization or use a NewSyncX(configuration) mechanism.

Do you mind explaining this last point

File: pkg/sync/isync.go
Comment (james-milligan) [reply pkg/sync/isync.go:61]: Another option is to map between the 'general' configuration structure and a sync specific one in the NewXYZ(cofiguration) methods, but this option feels inappropriate

**Comment by Kavindu-Dodan** [comment]
@toddbaert 

> Are you saying we should probably merge this (or something like it) before adding back TLS support to syncs?

Yes, we need to finalize how we obtain configurations (needs this for the cert path argument). One option is to introduce a dedicated flag and later add this flag to the mechanism this PR introduces (sync provider specific configurations). But then we need to agree whether **we will support both**: i.e - Expose startup option (ex:-  bearer token, tls cert path ...) & provider specific option of this PR.

p.s - I will avoid the configuration changes for now and work on the implementation level and open a draft PR 

@james-milligan 

> Do you mind explaining this last point, this PR implements the NewSyncX(configuration) option

Sorry if this was not clear. What I said was to introduce a **New** function to build the ISync struct and avoid having the configuration structure **embedded** in ISync impls - something like `NewK8sWithConfig`. This is to avoid having unrelevant configuration options (ex:- a nil bearer token inside the file ISync provider). Plus, we can enforce validations through **New** function (fail fast)

Plus, having such a contract helps to improve how we handle configurations. Let's say we use a common config object and use the `New` function. Given that we do not bind this configuration object to internals, our next improvement can focus on per sync-specific configuration object (instead of the global one). From ISync implementation point of view, there will be no change. We just need to have the parsing mechanism, create sync specific config object (i.e- K8sSyncConfig, fileSyncConfig, grpcSyncConfig) and simply introduce a `New` function to handle this.  

Anyway, this is my opinion 😉  open for opinions or discussions. I do not want to delay this PR further.

File: docs/configuration/flagd_start.md
Comment (toddbaert) [inline docs/configuration/flagd_start.md:24]: I could go either way, but I'm wondering if `sources` is better language for this. I think it comes down to if we want to think of `sync-providers` as more of an internal detail to flagd or not. Currently, I think we use `sources` language in OFO. I'm interested in everyone's thoughts.

File: cmd/start.go
Comment (toddbaert) [inline cmd/start.go:100]: Did we remove this deprecation?

File: cmd/start.go
Comment (toddbaert) [inline cmd/start.go:127]: The feedback we get here seems fine. When I mucked with the JSON the errors were clear enough.

File: cmd/start.go
Comment (toddbaert) [inline cmd/start.go:132]: The feedback we get here seems fine. When I mucked with the JSON the errors were clear enough.

File: cmd/start.go
Comment (toddbaert) [inline cmd/start.go:135]: The feedback we get here seems fine. When I mucked with the JSON the errors were clear enough.

File: pkg/runtime/from_config.go
Comment (toddbaert) [inline pkg/runtime/from_config.go:166]:
```suggestion
func SyncProviderArgValid(syncProviders string) ([]sync.ProviderConfig, error) {
```

File: pkg/runtime/from_config.go
Comment (toddbaert) [reply pkg/runtime/from_config.go:128]: I'll leave that up to you. I'd prefer to keep the codecov happy, but if some of the additions are troublesome to tests it's fine with me. I think you've got most of the important stuff in this PR.

**Review by toddbaert — APPROVED** [review]
Approved with minor feedback after local testing. I do think I like @Kavindu-Dodan 's idea [here](https://github.com/open-feature/flagd/pull/448#issuecomment-1456580414) though, and think it's worth consideration if you agree.... We could also do it in a new PR I guess.

File: cmd/start.go
Comment (james-milligan) [reply cmd/start.go:100]: reintroduced, nice catch

File: pkg/runtime/from_config.go
Comment (james-milligan) [reply pkg/runtime/from_config.go:164]: renamed to SyncProviderArgParse

File: docs/configuration/flagd_start.md
Comment (james-milligan) [reply docs/configuration/flagd_start.md]: ive introduced this change

File: cmd/start.go
Comment (Kavindu-Dodan) [inline cmd/start.go:70]: This must be removed right?

File: pkg/sync/http/http_sync.go
Comment (Kavindu-Dodan) [inline pkg/sync/http/http_sync.go:23]: Sorry for nitpicking, shall we separate **P**ublic and **p**rivate fields 

```suggestion
	URI         string
	Client      Client
	Cron        Cron
	LastBodySHA string
	Logger      *logger.Logger
	BearerToken string
	
	ready       bool
```

File: pkg/sync/kubernetes/kubernetes_sync.go
Comment (Kavindu-Dodan) [inline pkg/sync/kubernetes/kubernetes_sync.go:31]: Same with ordering 🥺  ?

**Comment by Kavindu-Dodan** [comment]
Overal looks good 👍

File: docs/configuration/configuration.md
Comment (Kavindu-Dodan) [inline docs/configuration/configuration.md:55]: Not sure if this is only my machine (Mac zsh). I was able to use the start command simply without escaping. Used single quotes 🤔 

ex:- 
`go run main.go start --sources='[{"uri":"config/samples/example_flags.json","provider":"file"}]' --debug
`

File: docs/configuration/configuration.md
Comment (beeme1mr) [inline docs/configuration/configuration.md:42]: Should we call them `SourceConfig` objects?

File: docs/configuration/configuration.md
Comment (beeme1mr) [inline docs/configuration/configuration.md:51]: What would happen if the `file` uri prefix was included? Can you give more examples below? I'm particularly interested in how the HTTP source config should look but including one for gRPC and Kubernetes would be useful as well.

File: docs/configuration/configuration.md
Comment (beeme1mr) [reply docs/configuration/configuration.md:55]: single quotes worked for me in Ubuntu on WSL2.

File: cmd/start.go
Comment (beeme1mr) [inline cmd/start.go:119]: I'm seeing this warning even when I don't pass `sync-provider-args`. Could you please double-check the logic? I would expect to only see this when it's actually being used.

**Review by beeme1mr — COMMENTED** [review]
It's probably unrelated to this change but I'm seeing two WRITE events in the logs every time a flag configuration changes using the file source. Could you please confirm that this change isn't the root cause. If it's not, please create a follow up issue to investigate.

File: cmd/start.go
Comment (james-milligan) [reply cmd/start.go:119]: switched to a string check

File: docs/configuration/configuration.md
Comment (james-milligan) [reply docs/configuration/configuration.md]: Added logic to parse `file:` and `core.openfeature.dev` prefixes, and added a reference to this in the docs

**Comment by james-milligan** [comment]
> It's probably unrelated to this change but I'm seeing two WRITE events in the logs every time a flag configuration changes using the file source. Could you please confirm that this change isn't the root cause. If it's not, please create a follow up issue to investigate.

Please can you provide more info - im not able to replicate this behaviour:

```
2023-03-08T09:27:41.069Z        debug   file/filepath_sync.go:121       Configuration config/samples/example_flags.json:  ALL   {"component": "sync", "sync": "filepath"}
2023-03-08T09:28:31.740Z        info    file/filepath_sync.go:78        filepath event: config/samples/example_flags.json CHMOD {"component": "sync", "sync": "filepath"}
2023-03-08T09:28:31.741Z        info    file/filepath_sync.go:78        filepath event: config/samples/example_flags.json WRITE {"component": "sync", "sync": "filepath"}
2023-03-08T09:28:31.741Z        debug   file/filepath_sync.go:121       Configuration config/samples/example_flags.json:  ALL   {"component": "sync", "sync": "filepath"}
2023-03-08T09:28:36.256Z        info    file/filepath_sync.go:78        filepath event: config/samples/example_flags.json CHMOD {"component": "sync", "sync": "filepath"}
2023-03-08T09:28:36.256Z        info    file/filepath_sync.go:78        filepath event: config/samples/example_flags.json WRITE {"component": "sync", "sync": "filepath"}
2023-03-08T09:28:36.256Z        debug   file/filepath_sync.go:121       Configuration config/samples/example_flags.json:  ALL   {"component": "sync", "sync": "filepath"}
```

I get repeated patterns of `CHMOD`, `WRITE` then a configuration `ALL` event using this startup command:
```
go run main.go start --sources='[{"uri":"config/samples/example_flags.json","provider":"file"}]' --debug 
```

File: cmd/start.go
Comment (skyerus) [inline cmd/start.go:65]:
```suggestion
			"2 required fields, uri (string) and provider (string). Documentation for this object: "+
```

File: pkg/sync/kubernetes/kubernetes_sync.go
Comment (skyerus) [inline pkg/sync/kubernetes/kubernetes_sync.go:32]: Is this definitely needed? Thought it was removed recently

File: pkg/sync/file/filepath_sync.go
Comment (skyerus) [inline pkg/sync/file/filepath_sync.go:22]: Needed?

File: cmd/start.go
Comment (james-milligan) [reply cmd/start.go:70]: deprecated :)

**Comment by toddbaert** [comment]
> > It's probably unrelated to this change but I'm seeing two WRITE events in the logs every time a flag configuration changes using the file source. Could you please confirm that this change isn't the root cause. If it's not, please create a follow up issue to investigate.
> 
> Please can you provide more info - im not able to replicate this behaviour:
> 
> ```
> 2023-03-08T09:27:41.069Z        debug   file/filepath_sync.go:121       Configuration config/samples/example_flags.json:  ALL   {"component": "sync", "sync": "filepath"}
> 2023-03-08T09:28:31.740Z        info    file/filepath_sync.go:78        filepath event: config/samples/example_flags.json CHMOD {"component": "sync", "sync": "filepath"}
> 2023-03-08T09:28:31.741Z        info    file/filepath_sync.go:78        filepath event: config/samples/example_flags.json WRITE {"component": "sync", "sync": "filepath"}
> 2023-03-08T09:28:31.741Z        debug   file/filepath_sync.go:121       Configuration config/samples/example_flags.json:  ALL   {"component": "sync", "sync": "filepath"}
> 2023-03-08T09:28:36.256Z        info    file/filepath_sync.go:78        filepath event: config/samples/example_flags.json CHMOD {"component": "sync", "sync": "filepath"}
> 2023-03-08T09:28:36.256Z        info    file/filepath_sync.go:78        filepath event: config/samples/example_flags.json WRITE {"component": "sync", "sync": "filepath"}
> 2023-03-08T09:28:36.256Z        debug   file/filepath_sync.go:121       Configuration config/samples/example_flags.json:  ALL   {"component": "sync", "sync": "filepath"}
> ```
> 
> I get repeated patterns of `CHMOD`, `WRITE` then a configuration `ALL` event using this startup command:
> 
> ```
> go run main.go start --sources='[{"uri":"config/samples/example_flags.json","provider":"file"}]' --debug 
> ```

I think this is all old behavior, and not much can be done about it. It seems to be all about how particular editors and programs update files. Even in the most basic cases, programs actually remove/write the file. The only exception is if the file being watched is a symlink and that symlink is updated. That's the only thing that seems to be completely atomic.

Even if I run something like: `sed -i 's/"defaultVariant": "on"/"defaultVariant": "off"/g' config/samples/example_flags.json` which uses `sed`'s "in-place" option (`-i`), the file is actually removed and replaced with new contents (I had not idea it worked that way, but [apparently it does](https://github.com/fsnotify/fsnotify/issues/143))

**Review by toddbaert — APPROVED** [review]
Re-tested and re-approved.

**Review by skyerus — APPROVED** [review]

**Pull request merged**
